truffleruby/src/main/c/cext/ruby.c

```diff
@@ -1872,13 +1872,19 @@ void rb_exc_raise(VALUE exception) {
 }
 
 VALUE rb_protect(VALUE (*function)(VALUE), VALUE data, int *status) {
-  // TODO CS 23-Jul-16
-  return function(data);
+  VALUE ary;
+  if (rb_block_given_p()) {
+    ary = truffle_invoke(RUBY_CEXT, "rb_protect_with_block", truffle_address_to_function(function), data, rb_block_proc());
+  } else {
+    ary = truffle_invoke(RUBY_CEXT, "rb_protect", truffle_address_to_function(function), data);
+  }    
```

Comment:
I would just use `rb_block_proc()` here and no branching, it returns `nil` if no block is found.